src/weight_windows.cpp

```diff
@@ -246,24 +349,360 @@ WeightWindow WeightWindows::get_weight_window(const Particle& p) const
+  // value will be ignored
+  xt::filter(new_bounds, sum <= 0.0).fill(-1.0);
+
+  // make sure the weight windows are ignored for any locations where the
+  // relative error is higher than the specified relative error threshold
+  xt::filter(new_bounds, rel_err > threshold).fill(-1.0);
+
+  // update the bounds of this weight window class
+  // noalias avoids additional memory allocation
+  xt::noalias(upper_ww_) = ratio * lower_ww_;
+}
+
+void WeightWindows::check_tally_update_compatibility(const Tally* tally)
+{
+  // define the set of allowed filters for the tally
+  const std::set<FilterType> allowed_filters = {
+    FilterType::MESH, FilterType::ENERGY, FilterType::PARTICLE};
+
+  // retrieve a mapping of filter type to filter index for the tally
+  auto filter_indices = tally->filter_indices();
+
+  // a mesh filter is required for a tally used to update weight windows
+  if (!filter_indices.count(FilterType::MESH)) {
+    fatal_error(
+      "A mesh filter is required for a tally to update weight window bounds");
+  }
+
+  // ensure the mesh filter is using the same mesh as this weight window object
+  auto mesh_filter = tally->get_filter<MeshFilter>();
+
+  // make sure that all of the filters present on the tally are allowed
+  for (auto filter_pair : filter_indices) {
+    if (allowed_filters.find(filter_pair.first) == allowed_filters.end()) {
+      fatal_error(fmt::format("Invalid filter type '{}' found on tally "
+                              "used for weight window generation.",
+        model::tally_filters[filter_pair.second]->type_str()));
+    }
+  }
+
+  if (mesh_filter->mesh() != mesh_idx_) {
+    int32_t mesh_filter_id = model::meshes[mesh_filter->mesh()]->id();
+    int32_t ww_mesh_id = model::meshes[this->mesh_idx_]->id();
+    fatal_error(fmt::format("Mesh filter {} uses a different mesh ({}) than "
+                            "weight window {} mesh ({})",
+      mesh_filter->id(), mesh_filter_id, id_, ww_mesh_id));
+  }
+
+  // if an energy filter exists, make sure the energy grid matches that of this
+  // weight window object
+  if (auto energy_filter = tally->get_filter<EnergyFilter>()) {
+    std::vector<double> filter_bins = energy_filter->bins();
+    std::set<double> filter_e_bounds(
+      energy_filter->bins().begin(), energy_filter->bins().end());
+    if (filter_e_bounds.size() != energy_bounds().size()) {
+      fatal_error(
+        fmt::format("Energy filter {} does not have the same number of energy "
+                    "bounds ({}) as weight window object {} ({})",
+          energy_filter->id(), filter_e_bounds.size(), id_,
+          energy_bounds().size()));
+    }
+
+    for (auto e : energy_bounds()) {
+      if (filter_e_bounds.count(e) == 0) {
+        fatal_error(fmt::format(
+          "Energy bounds of filter {} and weight windows {} do not match",
+          energy_filter->id(), id_));
+      }
+    }
+  }
+}
+
+void WeightWindows::export_to_hdf5(const std::string& filename) const
```

Comment:
Though we don't use it right now, being able to export them one-by-one is appealing to me. With `openmc_weight_windows_export` it's all or nothing. Do you think it would be useful to keep this method and expose it through the `openmc.lib` module? 

I've been toying around with the idea of using one set of weight windows to refine another or nested sets of weight windows to determine where they're most effective spatially, with one set of weight windows being the final useful product of the simulation. In case, being able to do something like:

```python
# run transport loop
ww1 = openmc.lib.WeightWindows.from_tally(tally1)
ww2 = openmc.lib.WeightWindows.from_tally(tally2) 
for i in range(iterations):
    openmc.lib.run()

    ww1.update_magic(tally1)
    ww2.update_magic(tally2)

# only want to keep ww2
ww2.to_hdf5("target-local-wws.h5")
```

Reply:
I would prefer that we don't have dead code that it is not being used -- if you decide to add this capability (as a follow-on PR?), you can add it back in at that time.

Reply:
Sounds good. Removing now.